source/hpack/decoder.d

```diff
@@ -0,0 +1,278 @@
+module HPACK.decoder;
+
+import HPACK.exception;
+import HPACK.huffman;
+import HPACK.tables;
+import HPACK.util;
+
+import std.range; // Decoder
+import std.bitmanip; // prefix encoding / decoding
+
+/** Module to implement an header decoder consistent with HPACK specifications (RFC 7541)
+  * The detailed description of the decoding process, examples and binary format details can
+  * be found at:
+  * Section 3: https://tools.ietf.org/html/rfc7541#section-3
+  * Section 6: https://tools.ietf.org/html/rfc7541#section-6
+  * Appendix C: https://tools.ietf.org/html/rfc7541#appendix-C
+*/
+alias HTTP2SettingValue = uint;
+
+/** implements an input range to decode an header block
+  * m_table is a reference to the original table
+  */
+struct HeaderDecoder(T = ubyte[])
+		if (isInputRange!T && (is(ElementType!T : char) || (is(ElementType!T : ubyte))))
+{
+	private {
+		ubyte[] m_range;
+		IndexingTable m_table; // only for retrieving data
+		HTTP2HeaderTableField[] m_decoded;
+		HTTP2HeaderTableField[] m_index; // to be appended
+		HTTP2HeaderTableField[] m_noindex;
+	}
+
+	this(T range, IndexingTable table) @trusted
+	{
+		static if(is(ElementType!T : char)) m_range = cast(ubyte[])range;
+		else m_range = range;
+		m_table = table;
+
+		decode();
+	}
+
+// InputRange specific methods
+	@property bool empty() @safe { return m_decoded.empty; }
+
+	@property HTTP2HeaderTableField front() @safe { return m_decoded.front; }
+
+	@property HTTP2HeaderTableField back() @safe { return m_decoded.back; }
+
+	void popFront() @trusted
+	{
+		enforce!HPACKException(!empty, "Cannot call popFront on an empty HeaderDecoder");
+
+		m_decoded.popFront();
+
+		// advance if data is still available
+		if(!m_range.empty) decode();
+	}
+
+	@property HTTP2HeaderTableField opIndex(int idx) @safe
+	{
+		assert(idx < m_decoded.length, "Invalid decoder index");
+		return m_decoded[idx];
+	}
+
+	@property HTTP2HeaderTableField[] opSlice(int start, int end) @safe
+	{
+		assert(start >= 0 && end < m_decoded.length, "Invalid decoder slice");
+		return m_decoded[start..end];
+	}
+
+	void put(T)(T range) @safe
+	{
+		static if(is(ElementType!T : char)) m_range ~= cast(ubyte[])range;
+		else m_range ~= range;
+
+		decode();
+	}
+
+	@property HTTP2HeaderTableField[] toIndex() @safe { return m_index; }
+
+	@property HTTP2HeaderTableField[] neverIndexed() @safe { return m_noindex; }
+
+// decoding
+	private {
+
+		void decode() @trusted
+		{
+			while(!m_range.empty) {
+
+				auto bbuf = [m_range[0]].toBitArray();
```

Comment:
Using `m_range[0 .. 1].toBitArray()` instead eliminates an extra GC allocation per byte here. That would just require `toBitArray()` to be rewritten so that it takes an array of `T` instead of `T` and passes the array to the `void[]` based constructor of `BitArray`.